### ISSUE TITLE
Coupons?

### ISSUE THREAD

**Issue by skylarmt** [jekkos/opensourcepos#477]
Is it possible to make barcodes that give a percent discount on a transaction?

If it isn't, this should be added.

**Comment by jzonetan** [comment]
yeah i was wondering that instead using the giftcard why not use coupon more flexible

**Comment by daN4cat** [comment]
Sure, free to contribute the feature. We need help.

**Comment by jekkos** [comment]
Maybe a barcode with giftcard number would be good here.. Also you can use giftcards without linking a customer which might already fit your requirements for the 'coupon'

**Comment by skylarmt** [comment]
The problem with using gift cards is it's impossible to print 50 coupons with the same barcode, and it's impossible to have a percent discount in one.  Gift cards would only be feasible if you only need a few coupons, and they are for a flat amount.

**Comment by jekkos** [comment]
So you'd probably want to extend giftcards and select their 'type' to give 

* fixed discount or percentage
* eventual link with items to which they can be applied

**Comment by skylarmt** [comment]
It looks like items can have a negative price, causing a discount on the whole sale.  Maybe it would be better to extend items with a percentage option instead of extending gift cards?

**Comment by swtrpeter1** [comment]
I would really like to be able to make coupons ! That would be great ! I would love to do 10% off an entire order or some promotional things like that. Someone please add this in !

**Comment by jimktrains** [comment]
What parameters would we be looking for with coupons?

* coupon code
* max uses
* percent or flat value and amount
* valid_after and valid_before
* specific items or the whole order
* can combine with other coupons / discounts
* max amount to discount
* min and max subtotals
* title

**Comment by daN4cat** [comment]
@jimktrains Thanks for the suggestions. I renumbered the bullet points so it's easier to respond:

1. yes
2. yes
3. yes
4. validity window
5. yes
6. could be but usually it's not cumulative. So it's a nice to have
7. yes
8. not sure
9. of course a title

and more...

10. barcode support
11. scan from the register and get it registered
12. printing or emails or text message (coupon code / 
13. attach to receipt
14. means to define a segment (group of people)
15. rules to suggest the coupon (after one sale, due to smaller purchase, etc.)

I worked for a Loyalty card company for a while and it can get really tricky, especially on the recommendation / allocation rules and segment identification.

Willing to contribute the feature?

**Comment by jimktrains** [comment]
4, exactly.

10 - 15, yeah :) I was thinking more in terms of rules when I wrote that, but yeah a bunch of admin stuff around them.

16. Apply automatically to sales.

I could potentially be able to take it on as part of the modifications I'm making to OSPOS for internal use. Right now we're evaluating it to see if it is what our store managers are looking for.  I'm not yet at that point and we havn't even talked about pushing anything back yet (it's in the back of all our minds). (I also have a bunch of other changes I'd like to try to get in as well if we go down this route.) I can get back to you in a few weeks with what direction we're going in.

**Comment by jimktrains** [comment]
I'm pushing for it! I anticipate some things will make it back. Much of what I'm doing isn't applicable  for the core project, e.g. handling Credit Card transactions and integration with our application, but the rest should/might be. I'll keep you posted :)

**Comment by daN4cat** [comment]
Out of curiosity, how do you plan to integrate with the Credit Card payment system?

**Comment by jimktrains** [comment]
We're already inside a PCI environment, so we simply grab Track 2 with a mag strip reader, send it to our processor. Alternatively you can type the card out and do a card not present transaction.

I modified the payment area on the register view to have sections for different methods. (Also helped with getting gift cards that can be under max utilized.) So that's where the fields for the card live. On submit, those are sent to our processor and we get a token back. The token is placed into a hidden field in the form and then the form is submitted to OSPOS.  

Inside OSPOS I do an authorization for the amount that was added to the card, or return an error if the authorization fails.  When a card is removed, I do a refund to release the authorization. If the sale is abandoned, the authorization will be released automatically by the bank after a certain amount of time (usually a day). (It's like a 2-phase commit in essence.) 

When the sale is completed, those authorizations are captured. I store the charge id and the last 4 of the card in the database in the sales_payments table. I also add the last 4 of the card to the payment_type so that they show up individually and can be removed and seen as such.

One of our goals is to move to having real terminals capable of EMV and contactless.  Those would have a service running on the local PC that we'd talk to through websockets. So instead of grabbing track 2 or entering details, selecting a credit card payment type would activate javascript that would talk to the terminal and eventually return a charge id that will get sent to OSPOS to be stored and on sale completion captured.

**Comment by daN4cat** [comment]
Sounds good, thanks for sharing. I wanted to find the time to investigate more the latter approach from your description. But currently my wife's shop uses and external PDQ machine connected to the Internet and so not sure there is at least a way to interact to push total and get back a payment ok/failure to automate the receipt issuing process.
Not sure if you are aware of such option or you really need to run all the payment module on your pos terminal?

**Comment by daN4cat** [comment]
And not sure that implies something extra in the PCI compliancy... that would need investigation too.

**Comment by jimktrains** [comment]
Because we currently accept cards where the clerk doesn't need to use a separate terminal, we want to keep a similar flow.

Having a fully separate machine is the easiest from a pci standpoint.

The back-and-forth is about the same, because the terminal doesn't give the application any credit card details. The back-and-forth just allows the application to have some control over the transaction. Each terminal vendor/gateway combo is different, though, so this would be difficult to have a generic solution for. Each vendor would require their own driver.

(Both would be under SAQ-B IP iirc). 

Accepting cards in software using something like Stripe to handle the actual card number is a little more complicated. SAQ-A EP isn't so bad, though, and attainable for someone familiar enough to install ospos.

Accepting card details onto your server can be a bigger headache. SAQ-D can be a big pain.

The thing is that most of the issues around pci aren't technical. Some are, shared hosting can be a potential issue, but at least ½ are business controls and internal training requirments. Most of the remaining technical issues would be specific to the installation, and not something we can help with in code.

If this is something the project is debating, my suggestion would be to see about incorporating something like Stripe elements, or otherwise using stripe.js. Out of the box SAQ-A EP is manageable for most people. However, you'll get. Card not present rates, which can be many millage points above card present transactions. Card present requires showing Stripe an AoC (attestation of compliance), but we can provide a config switch to use the API once enabled.

While very practical, something about only supporting a single vendor by default feels off. Not nessecarilly bad, just off. I guess if an eye is kept to build a framework, rather than just shoving it in there, it would be easier to add others of have people build custom ones for themselves.

Sorry, I talk a lot. Feel free to email me or hit me up in a chat if you'd like to discuss some of this outside this thread.

**Comment by daN4cat** [comment]
Thank you for the tips. I'll have a look into it when I find the time.

**Comment by jimktrains** [comment]
I'm still working on hammering out the complexities of various coupon types and how they fit into the flow of the various totals calculations. I have a feeling I'll end up rewriting large portions of Sales_lib to clean things up and better document how various things are related to each other.

**Comment by daN4cat** [comment]
I guess there is quite some work, it's not an easy addition and many use cases.

**Comment by jimktrains** [comment]
@daN4cat precisely. There are so many types of coupons and restrictions about how they can be used together that it becomes quite the thing to even specify.

My goal is to figure out how to specify those restrictions and then work out a structure for calculating prices and displaying which discounts are in effect per item.

From there I'll get it working for us (we need it soon) and then present what I've done and how it works here. If you feel it's a good path/a path you want to take, I can work on building out a PR for it.

**Comment by jimktrains** [comment]
@daN4cat Just a quick update.

Because so much of `Sale_lib` is based on the idea that a price is  specified by a (unit price, quantity, discount percent) triple, I've ended up having to rewrite giant chunks of it to support discounts in all the forms we support.

Is this something you'd be interested in having committed back? If so,  I can start cleaning up a branch to show and discuss. (We have a lot of specific changes just for our installation and I'll have to remove those.)

I've settled on a setup that is represented like this:

`coupons` table:

* `type` enum('automatic', `coupon') - automatic is applied, well, automatically and a coupon is applied by entering a code
* `discount_type` enum('flat', 'percent', 'product_price') - `product_price` pulls a price from `coupon_items`
* `discount_amount`
* `max_discount_value` - ceiling for the amount this coupon can discount
* `subtotal_threshold` - subtotal to qualify for the discount
* `starts_at`
* `ends_at`
* `coupon_code`
* `max_uses` - max times this coupon can be used
* `uses` - number of times it's been used
* `deleted`

`coupon_items`:

* `item_id`
* `coupon_id`
* `product_stock_floor` - sell at this price until there are x items remaining
* `max_per_order` - max items that can be purchased at this price
* `bonus_points`
* `price`

Oddly, we don't do BOGO or buy 2 get 1 sales? I have some ideas on how to work that into this system, though.

I've also made a bunch of smaller improvements I'll try to break out into separate PRs, but almost everything feels dependent on the changes to the `Sale_lib`.

One of the big ones is making historical receipts pull all data from the `sales_*` tables and not computing anything. I've noticed a bunch of small inconsistencies because of this. However, this includes a bunch of db changes as well.  Again, if you're interested, I can try to clean it up and post a PR.

**Comment by WebShells** [comment]
Sniffing interesting additions coming soon?  😎

**Comment by jimktrains** [comment]
@WebShells That depends on if the organization is OK with a rewrite of the `Sale_lib` :) It's a pretty big change, so I wanted to make sure it's even something that'd be considered before I go through and tease out things specific to us.

**Comment by jimktrains** [comment]
Being an ugly America, things I will definitely need help testing are where cash and totals aren't at the same precision/rounding and VAT.

Also, we manage promos and coupons through our system, so I don't really have an admin interface built out yet.

**Comment by daN4cat** [comment]
Rounding, precision and VAT. Sounds scary.... spent a lot to test and tune. We would need to build a test harness before we can make any change.

**Comment by jimktrains** [comment]
@daN4cat yes, I agree. I had gotten ci-phpunit-test working, but it's dependent upon some of our internal setup for the data for test-cases. It could be useful and I could build some more, truer unit tests out with it.

In terms of rounding in general, I'm making sure that it's done in an internally consistent way. I'm not sure what the cash rounding is called; we don't do it in the US. (We still use pennies, the way God intended! (Even if they cost more than a penny to mint :()) 

We also don't do VAT, so I'm not exactly sure how it works or if I've correctly used the existing tax code.

I'll try to get this cleaned up and make a PR for it. I don't think the issues are quite as bad as I may be making them sound. I just want to call out known things that will need an even closer eye then everything else.

**Comment by jekkos** [comment]
@daN4cat was thinking exactly the same there. When having no tests we rely on users accidentally creating bug reports. If we would have had some tests then people wanting to contribute would directly have a feel on whether they are causing regressions or not..

**Comment by jekkos** [comment]
Adding a test harness should be as easy as `composer require kenjis/ci-phpunit-test --dev`

would be nice if we don't need to commit that to the repo, and then have the build install it and run our the php files we add as tests themselves

**Comment by jekkos** [comment]
Adding a test harness should be as easy as `composer require kenjis/ci-phpunit-test --dev`

would be nice if we don't need to commit that to the repo, and then have the build install it and run our the php files we add as tests themselves

**Comment by jimktrains** [comment]
It seems to generate a lot of files. We might be able to not have to have all of them committed.

I did have to make some changes to allow it to bypass/pass csrf protections when testing and mocking http requests.

I can take this and get a basic test case working on a branch off of master. I can then take that and base all my changes and tests off of it. Would that work for you guys?

**Comment by daN4cat** [comment]
Would be good if you could contribute some tests so we get this kick started and improve. We really need this...

**Comment by jekkos** [comment]
Sounds great @jimktrains ! I'd really like to see this part of the codebase improved.

**Comment by daN4cat** [comment]
@jimktrains just wondering if you got somewhere with the test part?

**Comment by jimktrains** [comment]
@dan4cat I got the framework set up and started writing tests, but I've basically been moved to another project as my main focus, so it's going slowly. I can see if I think what I have is good to build on and try to push it and make a PR.

I'm also still in the middle of the thr coupon rewrite for us, but it's not my main priority at the moment:(

**Comment by daN4cat** [comment]
Maybe you can drop a first draft of the tests and see if we can add on top of that.

**Comment by DamianoSilverhand** [comment]
> @daN4cat I got the framework set up and started writing tests, but I've basically been moved to another project as my main focus, so it's going slowly. I can see if I think what I have is good to build on and try to push it and make a PR.
> 
> I'm also still in the middle of the the coupon rewrite for us, but it's not my main priority at the moment:(

@jimktrains How did the Coupons implementation feature go? we need something similar for our local setup and will appreciate some hints and feedback

**Comment by jimktrains** [comment]
@DamianoSilverhand I'm sorry it's taken so long to respond to you. I will write up something about what I did. It's been working well for us.

I also want to apologize to the community as a whole. I wasn't able to get the time I needed to merge from the public master and clean up the implementation in order to commit it back.  I  will try again to get a little time to, but I can't promise a timeline unfortunately :disappointed:

**Comment by jimktrains** [comment]
@DamianoSilverhand The biggest thing I did was restructure how the cart was computed. I had a "Cart" model that had some methods and a list of "Cart_Item"s and a list of "Cart_Coupon"s.  The coupons would be sorted by the Cart (more on that later), then each Coupon would be passed the list of "Cart_Item"s. If the coupon's subtotal threshold was reached, the stock floor hadn't been reached, and the max uses has been exceeded, the Coupon would apply itself incrementally to each Item in the cart, if it was applicable to that item, until it's applied its maximum discount it processed all of the Cart_Items.

Each Cart_Item stored the amount of the discount on itself. The Cart_Item also called out to the service we used to compute taxes and stored its own taxes. The Sales_lib would then use the information stored on the Cart_Items to compute the total discounts, taxes, subtotals, &c. Otherwise it was just a matter of figuring out how to redo some of the calculations as they are/were scattered all through the `Sales_lib`.

The cart would sort the Coupons using 

```php
usort($this->coupons, 'Cart_coupon::cmp_ordering');
```

in the Cart and `Cart_coupon::cmp_ordering` is

```php
  public static function cmp_ordering($ca, $cb)
  {
    $a = $ca->get_ordering_key();
    $b = $cb->get_ordering_key();

    $l = min(count($a), count($b));
    for ($i = 0; $i < $l; $i++)
    {
      if ($a[$i] != $b[$i])
      {
        if ($a[$i] < $b[$i])
        {
          return -1;
        }
        return 1;
      }
    }
    return 0;
  }

  public function get_ordering_key()
  {
    // Apply price discounts and then rewards points
    $benefit            = $this->is_discount_coupon() ? '1-discount' : '2-rewards-points';

    // Apply all automatic coupons first.
    $type               = $this->is_type_automatic() ? '1-automatic' : '2-manual';

    // Apply coupons constrained to specific products before apply ones
    // for the whole order.
    $applies_to         = $this->has_item_constraints() ? '1-item' : '2-order';

    // Apply all coupons that don't have a subtotal threshhold to meet before
    // ones that have a threshold.
    $conditional        = $this->is_unconditional() ? '1-unconditional' : "2-conditional";

    // Apply coupons that change the item price before coupons that represent a
    // discount.
    $application_method = '999-none';
    if ($this->is_discount_type_product_price()) {
      $application_method = '1-item_price';
    } elseif ($this->is_discount_type_percent()) {
      $application_method = '2-percent';
    } elseif ($this->is_discount_type_flat()) {
      $application_method = '3-flat';
    }

    $ordering_key = [$benefit, $type, $applies_to, $application_method, $conditional];
    return $ordering_key;
  }
```

One of the more difficult part of the whole implementation was getting that ordering to work in a way that I wanted. 

My `ospos_coupons` table is actually a view against our ecommerce coupons, so I can't give you a specific `create table` statement, but this is what the table looks like

```
+---------------------+----------------------+------+-----+---------+-------+
| Field               | Type                 | Null | Key | Default | Extra |
+---------------------+----------------------+------+-----+---------+-------+
| id                  | int(11)              | NO   |     | 0       |       |
| coupon_id           | int(11)              | NO   |     | NULL    |       |
| item_id             | int(11)              | YES  |     | NULL    |       |
| product_stock_floor | smallint(5) unsigned | NO   |     | 0       |       |
| max_per_order       | int(10) unsigned     | NO   |     | 1       |       |
| bonus_points        | int(1)               | NO   |     | 0       |       |
| price               | double(10,2)         | NO   |     | 0.00    |       |
+---------------------+----------------------+------+-----+---------+-------+

```

Ditto for the `ospos_coupon_products`

```
+---------------------+----------------------+------+-----+---------+-------+
| Field               | Type                 | Null | Key | Default | Extra |
+---------------------+----------------------+------+-----+---------+-------+
| id                  | int(11)              | NO   |     | 0       |       |
| coupon_id           | int(11)              | NO   |     | NULL    |       |
| item_id             | int(11)              | YES  |     | NULL    |       |
| product_stock_floor | smallint(5) unsigned | NO   |     | 0       |       |
| max_per_order       | int(10) unsigned     | NO   |     | 1       |       |
| bonus_points        | int(1)               | NO   |     | 0       |       |
| price               | double(10,2)         | NO   |     | 0.00    |       |
+---------------------+----------------------+------+-----+---------+-------+
```

I added an application setting for coupons that'd be added to every cart representing automatic discounts and such. Those coupons could be tied to specific items or just be a general discount.

Feel free to ask if you have any other questions. If I remember anything else substantial I'll reply here.


I'll see if I can talk to my boss and see if I can get some time during work or permission to port the code after work to the current main branch.
